modules/bmp/image_loader_bmp.cpp

```diff
@@ -0,0 +1,190 @@
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "image_loader_bmp.h"
+
+Error ImageLoaderBMP::convert_to_image(Ref<Image> p_image,
+                                       const uint8_t *p_buffer,
+                                       const uint8_t *p_color_buffer,
+                                       const bmp_header_s &p_header) {
+  Error err = OK;
+
+  if (p_buffer == NULL)
+    err = FAILED;
+
+  if (err == OK) {
+    size_t index = 0;
+    size_t width =
+        static_cast<size_t>(p_header.bmp_info_header.bmp_width < 0
+                                ? -p_header.bmp_info_header.bmp_width
+                                : p_header.bmp_info_header.bmp_width);
+    size_t height =
+        static_cast<size_t>(p_header.bmp_info_header.bmp_height < 0
+                                ? -p_header.bmp_info_header.bmp_height
+                                : p_header.bmp_info_header.bmp_height);
+    size_t bits_per_pixel =
+        static_cast<size_t>(p_header.bmp_info_header.bmp_bit_count);
+
+    uint32_t line_width = ((p_header.bmp_info_header.bmp_width *
+                            p_header.bmp_info_header.bmp_bit_count / 8) +
+                           3) &
+                          ~3;
+
+    PoolVector<uint8_t> image_data;
+    err = image_data.resize(width * height * 4);
+
+    PoolVector<uint8_t>::Write image_data_w = image_data.write();
+    uint8_t *write_buffer = image_data_w.ptr();
+
+    if (p_header.bmp_info_header.bmp_compression == 0) {
+      const uint8_t *line = p_buffer + (line_width * (height - 1));
+      for (unsigned int i = 0; i < height; i++) {
+        const uint8_t *line_ptr = line;
+        for (unsigned int j = 0; j < width; j++) {
+          if (bits_per_pixel == 24) {
+            uint32_t color = *((uint32_t *)line_ptr);
+
+            write_buffer[index + 2] = color & 0xff;
+            write_buffer[index + 1] = (color >> 8) & 0xff;
+            write_buffer[index + 0] = (color >> 16) & 0xff;
+            write_buffer[index + 3] = 0xff;
+            index += 4;
+            line_ptr += 3;
+          } else if (bits_per_pixel == 32) {
+            uint32_t color = *((uint32_t *)line_ptr);
+            write_buffer[index + 2] = color & 0xff;
+            write_buffer[index + 1] = (color >> 8) & 0xff;
+            write_buffer[index + 0] = (color >> 16) & 0xff;
+            write_buffer[index + 3] = color >> 24;
+            index += 4;
+            line_ptr += 4;
+          } else {
+            err = FAILED;
```

Comment:
Would it be better if we check this from the start, also with bit_per_pixel condition?

Reply:
Good spot, I've changed that now to have a more sensible error handler. Having a bit of trouble with the Clang precommit hooks though at the moment though...